### ISSUE TITLE
Failed demangling of Lambdas

### ISSUE THREAD

**Issue by mitsuhiko** [fitzgen/cpp_demangle#97]
This symbol:

```
_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv
```

Comes from

```
namespace {
	int hello() {
                ([] { printf("test"); })();
		return 0;
	}
}
```

Not sure what it should demangle to as c++filt also has problems with it.

**Comment by mstange** [comment]
With an extra underscore, `c++filt __ZN12_GLOBAL__N_15helloEv` gives `(anonymous namespace)::hello()`.

**Comment by mitsuhiko** [comment]
@mstange note that i originally copied the wrong identifier into the issue and edited it later. However I just realized that i have an ancient c++filt so most likely it works in new ones.

**Comment by mstange** [comment]
Oh, I see. The c++filt version I have isn't able to demangle `_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv` either, extra underscore or not.

**Comment by jan-auer** [comment]
FWIW, clang's own `abi::__cxa_demangle` (`cxxabi.h`) demangles it to `(anonymous namespace)::hello()::$_0::operator()() const`.

**Comment by jan-auer** [comment]
Here's a minimal example to reproduce:

```c++
#include <cxxabi.h>
#include <iostream>

using namespace std;

int main() {
  char *buf = (char *)malloc(1024);
  int status;
  size_t size;
  abi::__cxa_demangle("_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv", buf, &size, &status);
  cout << "status " << status << endl << buf << endl;
  return 0;
}
```

**Comment by fitzgen** [comment]
Thanks for the bug report!

FWIW, my binutils' `c++filt` will demangle the same as the llvm demangler:

```
$ c++filt --version
GNU c++filt (GNU Binutils for Ubuntu) 2.28
Copyright (C) 2017 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or (at your option) any later version.
This program has absolutely no warranty.

$ echo '_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv' | c++filt
(anonymous namespace)::hello()::$_0::operator()() const
```

**Comment by fitzgen** [comment]
Looks like we're failing to parse the symbol:

```
(MangledName "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
    (Encoding "ZN12_GLOBAL__N_15helloEvENK3$_0clEv"
        (Name "ZN12_GLOBAL__N_15helloEvENK3$_0clEv"
            (NestedName "ZN12_GLOBAL__N_15helloEvENK3$_0clEv"
            )
            (UnscopedName "ZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                (UnqualifiedName "ZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                    (OperatorName "ZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                        (SimpleOperatorName "ZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                        )
                    )
                    (CtorDtorName "ZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                    )
                    (SourceName "ZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                    )
                    (UnnamedTypeName "ZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                    )
                )
            )
            (UnscopedTemplateNameHandle "ZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                (UnscopedName "ZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                    (UnqualifiedName "ZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                        (OperatorName "ZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                            (SimpleOperatorName "ZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                            )
                        )
                        (CtorDtorName "ZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                        )
                        (SourceName "ZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                        )
                        (UnnamedTypeName "ZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                        )
                    )
                )
                (Substitution "ZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                    (WellKnownComponent "ZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                    )
                )
            )
            (LocalName "ZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                (Encoding "N12_GLOBAL__N_15helloEvENK3$_0clEv"
                    (Name "N12_GLOBAL__N_15helloEvENK3$_0clEv"
                        (NestedName "N12_GLOBAL__N_15helloEvENK3$_0clEv"
                            (CvQualifiers "12_GLOBAL__N_15helloEvENK3$_0clEv"
                            )
                            (RefQualifier "12_GLOBAL__N_15helloEvENK3$_0clEv"
                            )
                            (PrefixHandle "12_GLOBAL__N_15helloEvENK3$_0clEv"
                                (PrefixHandle iteration "12_GLOBAL__N_15helloEvENK3$_0clEv"
                                    (UnqualifiedName "12_GLOBAL__N_15helloEvENK3$_0clEv"
                                        (OperatorName "12_GLOBAL__N_15helloEvENK3$_0clEv"
                                            (SimpleOperatorName "12_GLOBAL__N_15helloEvENK3$_0clEv"
                                            )
                                        )
                                        (CtorDtorName "12_GLOBAL__N_15helloEvENK3$_0clEv"
                                        )
                                        (SourceName "12_GLOBAL__N_15helloEvENK3$_0clEv"
                                            (Identifier "_GLOBAL__N_1"
                                            )
                                        )
                                    )
SubstitutionTable::insert @ 0: Prefix(Unqualified(Source(SourceName(Identifier { start: 6, end: 18 }))))
                                )
                                (PrefixHandle iteration "5helloEvENK3$_0clEv"
                                    (SourceName "5helloEvENK3$_0clEv"
                                        (Identifier "hello"
                                        )
                                    )
SubstitutionTable::insert @ 1: Prefix(Nested(BackReference(0), Source(SourceName(Identifier { start: 19, end: 24 }))))
                                )
                                (PrefixHandle iteration "EvENK3$_0clEv"
                                )
                            )
SubstitutionTable::pop @ 2: Some(Prefix(Nested(BackReference(0), Source(SourceName(Identifier { start: 19, end: 24 })))))
                        )
                    )
                    (BareFunctionType "vENK3$_0clEv"
                        (TypeHandle "vENK3$_0clEv"
                            (BuiltinType "vENK3$_0clEv"
                                (StandardBuiltinType "vENK3$_0clEv"
                                )
                            )
                        )
                        (TypeHandle "ENK3$_0clEv"
                            (BuiltinType "ENK3$_0clEv"
                                (StandardBuiltinType "ENK3$_0clEv"
                                )
                            )
                            (ClassEnumType "ENK3$_0clEv"
                                (Name "ENK3$_0clEv"
                                    (NestedName "ENK3$_0clEv"
                                    )
                                    (UnscopedName "ENK3$_0clEv"
                                        (UnqualifiedName "ENK3$_0clEv"
                                            (OperatorName "ENK3$_0clEv"
                                                (SimpleOperatorName "ENK3$_0clEv"
                                                )
                                            )
                                            (CtorDtorName "ENK3$_0clEv"
                                            )
                                            (SourceName "ENK3$_0clEv"
                                            )
                                            (UnnamedTypeName "ENK3$_0clEv"
                                            )
                                        )
                                    )
                                    (UnscopedTemplateNameHandle "ENK3$_0clEv"
                                        (UnscopedName "ENK3$_0clEv"
                                            (UnqualifiedName "ENK3$_0clEv"
                                                (OperatorName "ENK3$_0clEv"
                                                    (SimpleOperatorName "ENK3$_0clEv"
                                                    )
                                                )
                                                (CtorDtorName "ENK3$_0clEv"
                                                )
                                                (SourceName "ENK3$_0clEv"
                                                )
                                                (UnnamedTypeName "ENK3$_0clEv"
                                                )
                                            )
                                        )
                                        (Substitution "ENK3$_0clEv"
                                            (WellKnownComponent "ENK3$_0clEv"
                                            )
                                        )
                                    )
                                    (LocalName "ENK3$_0clEv"
                                    )
                                )
                            )
                            (Substitution "ENK3$_0clEv"
                                (WellKnownComponent "ENK3$_0clEv"
                                )
                            )
                            (FunctionType "ENK3$_0clEv"
                                (CvQualifiers "ENK3$_0clEv"
                                )
                            )
                            (ArrayType "ENK3$_0clEv"
                            )
                            (PointerToMemberType "ENK3$_0clEv"
                            )
                            (TemplateParam "ENK3$_0clEv"
                            )
                            (TemplateTemplateParamHandle "ENK3$_0clEv"
                                (Substitution "ENK3$_0clEv"
                                    (WellKnownComponent "ENK3$_0clEv"
                                    )
                                )
                                (TemplateParam "ENK3$_0clEv"
                                )
                            )
                            (Decltype "ENK3$_0clEv"
                            )
                            (CvQualifiers "ENK3$_0clEv"
                            )
                        )
                    )
                )
                (Name "NK3$_0clEv"
                    (NestedName "NK3$_0clEv"
                        (CvQualifiers "K3$_0clEv"
                        )
                        (RefQualifier "3$_0clEv"
                        )
                        (PrefixHandle "3$_0clEv"
                            (PrefixHandle iteration "3$_0clEv"
                                (UnqualifiedName "3$_0clEv"
                                    (OperatorName "3$_0clEv"
                                        (SimpleOperatorName "3$_0clEv"
                                        )
                                    )
                                    (CtorDtorName "3$_0clEv"
                                    )
                                    (SourceName "3$_0clEv"
                                        (Identifier "$_0"
                                        )
                                    )
                                    (UnnamedTypeName "3$_0clEv"
                                    )
                                )
                            )
                        )
                    )
                    (UnscopedName "NK3$_0clEv"
                        (UnqualifiedName "NK3$_0clEv"
                            (OperatorName "NK3$_0clEv"
                                (SimpleOperatorName "NK3$_0clEv"
                                )
                            )
                            (CtorDtorName "NK3$_0clEv"
                            )
                            (SourceName "NK3$_0clEv"
                            )
                            (UnnamedTypeName "NK3$_0clEv"
                            )
                        )
                    )
                    (UnscopedTemplateNameHandle "NK3$_0clEv"
                        (UnscopedName "NK3$_0clEv"
                            (UnqualifiedName "NK3$_0clEv"
                                (OperatorName "NK3$_0clEv"
                                    (SimpleOperatorName "NK3$_0clEv"
                                    )
                                )
                                (CtorDtorName "NK3$_0clEv"
                                )
                                (SourceName "NK3$_0clEv"
                                )
                                (UnnamedTypeName "NK3$_0clEv"
                                )
                            )
                        )
                        (Substitution "NK3$_0clEv"
                            (WellKnownComponent "NK3$_0clEv"
                            )
                        )
                    )
                    (LocalName "NK3$_0clEv"
                    )
                )
            )
        )
        (SpecialName "ZN12_GLOBAL__N_15helloEvENK3$_0clEv"
        )
    )
    (TypeHandle "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
        (BuiltinType "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
            (StandardBuiltinType "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
            )
        )
        (ClassEnumType "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
            (Name "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                (NestedName "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                )
                (UnscopedName "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                    (UnqualifiedName "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                        (OperatorName "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                            (SimpleOperatorName "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                            )
                        )
                        (CtorDtorName "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                        )
                        (SourceName "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                        )
                        (UnnamedTypeName "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                        )
                    )
                )
                (UnscopedTemplateNameHandle "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                    (UnscopedName "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                        (UnqualifiedName "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                            (OperatorName "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                                (SimpleOperatorName "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                                )
                            )
                            (CtorDtorName "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                            )
                            (SourceName "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                            )
                            (UnnamedTypeName "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                            )
                        )
                    )
                    (Substitution "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                        (WellKnownComponent "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                        )
                    )
                )
                (LocalName "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                )
            )
        )
        (Substitution "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
            (WellKnownComponent "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
            )
        )
        (FunctionType "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
            (CvQualifiers "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
            )
        )
        (ArrayType "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
        )
        (PointerToMemberType "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
        )
        (TemplateParam "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
        )
        (TemplateTemplateParamHandle "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
            (Substitution "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                (WellKnownComponent "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
                )
            )
            (TemplateParam "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
            )
        )
        (Decltype "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
        )
        (CvQualifiers "_ZZN12_GLOBAL__N_15helloEvENK3$_0clEv"
        )
    )
)
```

**Comment by fitzgen** [comment]
I think its the `$`. If I replace the `$` with `z`, then we parse and format the symbol, but it isn't quite formatted correctly:

```
(anonymous namespace)::hello()()::z_0::operator() const
```

**Comment by fitzgen** [comment]
I have a fix for the `$` parsing, but don't have time to look into the formatting at the moment: https://github.com/fitzgen/cpp_demangle/pull/98

**Comment by fitzgen** [comment]
Fixed in 0.2.6
